Src/ILGPU/IR/Construction/BasicBlock.Builder.cs

```diff
@@ -511,13 +511,29 @@ public void ReplaceWithCall(Value value, Method implementationMethod)
             protected override T CreateTerminator<T>(T node)
             {
                 Terminator?.Replace(node);
-                return (Terminator = node).AsNotNullCast<T>();
+
+                Terminator = node;
+
+                switch (node.GetType().FullName)
```

Comment:
Is type of this concrete value important here or typeof(T) in general? If second, then I think it's better to use typeof(T)

Reply:
`System.Type` can't be used in `switch(X)`, so this would be the reason why. I agree that what you're suggesting would be more semantic, however, if it were allowed by the compiler.

Reply:
`typeof(T).FullName` can be used in a similar way too. 

It's just that node's type can be different from that of `typeof(T)` e.g.
```csharp
interface QQ {}
class WW : QQ {}

static class Test
{
    public static bool Z<T>(T value)
    {
        return typeof(T) == value.GetType();
    }
}

QQ ww = new WW();
Test.Z(ww) // false
```

This might be not a problem at all though

Reply:
I do in fact need to use `object.GetType()` in order for this type check to work because `T` will always be `IR.Values.TerminatorValue` for the erroneous calls in question.